### PR TITLE
Add dropout

### PR DESCRIPTION
**Pull request by xbelonogov**

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -2,17 +2,19 @@
...
         [
-            "youtokentome/cpp/yttm.cpp",
+            # "youtokentome/cpp/yttm.cpp",
+            "youtokentome/cpp/yttm.pyx",
             "youtokentome/cpp/bpe.cpp",
             "youtokentome/cpp/utils.cpp",
             "youtokentome/cpp/utf8.cpp",
         ],
-        extra_compile_args=["-std=c++11", "-pthread", "-O3"],
+        extra_compile_args=["-std=c++14", "-pthread", "-O3"],
```
Comment (yutkin) [inline setup.py]: Wheels build with C++11.

File: youtokentome/yttm_cli.py [hunk 2]
```diff
@@ -98,7 +98,10 @@ def bpe(data, model, vocab_size, coverage, n_threads, pad_id, unk_id, bos_id, eo
 @click.option(
     "--stream", is_flag=True, help="Process each line before reading the next one."
 )
-def encode(model, output_type, n_threads, bos, eos, reverse, stream):
+@click.option(
+    "--dropout_prob", type=click.FLOAT, default=0, show_default=True, help="Process each line before reading the next one."
```
Comment (kefirski) [inline youtokentome/yttm_cli.py]: it may be helpful to add the link on the paper (https://arxiv.org/abs/1910.13267) in help.

Comment (kefirski) [reply youtokentome/yttm_cli.py]: And it seems like current help contains unrelated information

Comment (kefirski) [reply youtokentome/yttm_cli.py]: Also, I would recommend adding callback on dropout option to throw an exception if the value is not within [0, 1] range
https://click.palletsprojects.com/en/7.x/options/#callbacks-for-validation

File: youtokentome/cpp/yttm.pyx [hunk 3]
```diff
@@ -84,38 +84,40 @@ cdef class BPE:
...
         cdef vector[string] s
         cdef vector[vector[string]] ret_subwords
         cdef vector[vector[int]] ret_ids
         cdef Status status
+        if dropout_prob < 0 or dropout_prob > 1:
+            raise ValueError("dropout_prob value must be in the range [0, 1]. Current value of dropout_prob = " + str(dropout_prob))
         if output_type == 'id':
             if isinstance(sentences, str):
                 s = [sentences.encode()]
                 assert len(s) == 1
```
Comment (kefirski) [inline youtokentome/cpp/yttm.pyx]: Could you please replace asserts without error information with `if` block, which throws an exception?

File: youtokentome/yttm_cli.py (hunk 2 above)
Comment (yutkin) [reply youtokentome/yttm_cli.py]: > Also, I would recommend adding callback on dropout option to throw an exception if the value is not within [0, 1] range

We already do it here https://github.com/VKCOM/YouTokenToMe/pull/45/files#diff-fe449cd06a471c4e41e1deae980a3badR92

File: youtokentome/cpp/yttm.pyx (hunk 3 above)
Comment (xbelonogov) [reply youtokentome/cpp/yttm.pyx]: Looks like this assert is always true. I removed it.

File: youtokentome/yttm_cli.py (hunk 2 above)
Comment (kefirski) [reply youtokentome/yttm_cli.py]: ok then

**Pull request merged**